sklearn/naive_bayes.py

```diff
@@ -1528,3 +1541,600 @@ def _joint_log_likelihood(self, X):
             jll += self.feature_log_prob_[i][:, indices].T
         total_ll = jll + self.class_log_prior_
         return total_ll
+
+
+def _fit_one(estimator, X, y, message_clsname="", message=None, **fit_params):
+    """Call ``estimator.fit`` and print elapsed time message.
+
+    See :func:`sklearn.pipeline._fit_one`.
+    """
+    with _print_elapsed_time(message_clsname, message):
+        return estimator.fit(X, y, **fit_params)
+
+
+def _partial_fit_one(estimator, X, y, message_clsname="", message=None, **fit_params):
+    """Call ``estimator.partial_fit`` and print elapsed time message.
+
+    See :func:`sklearn.pipeline._fit_one`.
+    """
+    with _print_elapsed_time(message_clsname, message):
+        return estimator.partial_fit(X, y, **fit_params)
+
+
+def _jll_one(estimator, X):
+    """Call ``estimator.predict_joint_log_proba``.
+
+    See :func:`sklearn.pipeline._transform_one`.
+    """
+    return estimator.predict_joint_log_proba(X)
+
+
+class ColumnwiseNB(_BaseNB, _BaseComposition):
+    """
+    Column-wise Naive Bayes meta-estimator.
+
+    This estimator combines various naive Bayes estimators by applying them
+    to different column subsets of the input and joining their predictions
+    according to the naive Bayes assumption. This is useful when features are
+    heterogeneous and follow different kinds of distributions.
+
+    Parameters
+    ----------
+    nb_estimators : list of tuples
```

Comment:
There were reasons for that, which may still be actual. Please see the preceding discussion:
- Item 14. of the top post
- https://github.com/scikit-learn/scikit-learn/pull/22574#pullrequestreview-922451649
- https://github.com/scikit-learn/scikit-learn/pull/22574#issuecomment-1080030147
- https://github.com/scikit-learn/scikit-learn/pull/22574#issuecomment-1094332808

P.S. Thanks for taking time to review, I will be able to go through all your comments within a day.

Reply:
I would really prefer to have `estimators` and check what is the reasoning behind the duck-typing of the tests. We did a recent refactoring for making all ensemble estimators switch from `base_estimator` to `estimator`. I think that we should avoid to have to do the same :)

Reply:
I am not sure this is exactly the code that was failing the new estimator before I renamed `estimators` into `nb_estimators`, but anyway this is an example of what I mean by "duck-typing":
https://github.com/scikit-learn/scikit-learn/blob/5201045fbe8423ae2d0e3d0465c023da3ed7fa85/sklearn/utils/estimator_checks.py#L411-L431
I could, of course, check `issubclass(_BaseNB)` and this would work. I wonder, was there a reason for not listing `StackingClassifier` and `VotingClassifier` classes explicitly? Or is it just being "pythonic"?


Now, let's look at the bigger picture for a moment. I am not suggesting anything specific, just thinking out loud. 
In sklearn there are:
- homogeneous ensembles with `estimator=est` parameter, takes a *single* instance; (formerly, `base_estimator`)
- heterogeneous ensembles with `estimators=[('name', est), ...]` parameter taking list of *pairs*
- `Pipeline` and `FeatureUnion`, which also take lists of *pairs* `steps=[('name', est), ...]` and `transformer_list=[('name', est), ...]`, respectively.
- `ColumnTranformer`, which takes a list of *triplets* `transformers=[('name', est, cols), ...]`
- and `ColumnwiseNB`, which also takes a list of *triplets* `nb_estimators=[('name', est, cols), ...]` or some other name?

Putting aside the issues I had due duck-typing in tests, is it aesthetically correct to re-use the name `estimators`?

Technically, `ColumnwiseNB` is an heterogeneous ensemble, but has some things in common with `ColumnTransformer` (which can be potentially factorised):
  - needs an adapter from triplets to pairs to make `set_params` work with things like `name__alpha`, see this [comment below](https://github.com/scikit-learn/scikit-learn/pull/22574#discussion_r1061135797)
  - selects columns, so needs to validate all `cols`. Also, column selection gets in the way of input validation. For callable and string-based selection we have to keep pandas DataFrame unconverted, but for all other cases it seems better to convert `X` to ndarray from the get-go. `ColumnTransformer` stores string-to-integer dictionary, I forgot why exactly, and `ColumnwiseNB` does not do that at the moment.
  - needs to check that the list contains triplets.

Reply:
Regarding various lists of tuples that pass estimators to meta-estimators etc. In addition to lists of pairs `[('name', est), ...]` and triplets `[('name', est, cols), ...]`, there is potentially a list of pairs `[(est, cols), ...]` or `[(est, col), ...]`, as in here #14253.